src/liballoc/collections/btree/map.rs

```diff
@@ -525,7 +569,39 @@ impl<K: Ord, V> BTreeMap<K, V> {
         }
     }
 
-    /// Clears the map, removing all values.
+    /// Clears the map, returning all key-value pairs as an iterator.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(btree_drain_retain)]
+    /// use std::collections::BTreeMap;
+    ///
+    /// let mut a = BTreeMap::new();
+    /// a.insert(1, "a");
+    /// a.insert(2, "b");
+    ///
+    /// for (k, v) in a.drain().take(1) {
+    ///     assert!(k == 1 || k == 2);
+    ///     assert!(v == "a" || v == "b");
+    /// }
+    ///
+    /// assert!(a.is_empty());
+    /// ```
+    #[unstable(feature = "btree_drain_retain", issue = "42849")]
+    pub fn drain(&mut self) -> IntoIter<K, V> {
+        let length = mem::replace(&mut self.length, 0);
+        let root = mem::replace(&mut self.root, node::Root::shared_empty_root());
+        let root1 = unsafe { ptr::read(&root).into_ref() };
+        let root2 = unsafe { ptr::read(&root).into_ref() };
+        let front = first_leaf_edge(root1);
+        let back = last_leaf_edge(root2);
+        IntoIter { front, back, length }
```

Comment:
Even simpler: `mem::replace(self, BTreeMap::new()).into_iter()`

Reply:
All right for the BTreeMap::new(), but isn't the `mem::forget(self);` in into_iter going to wreck this?

Reply:
No, `into_iter` is intentionally deconstructing the map, just as it always would. It will only affect the one you pull out of `self`, and the `new()` one will remain untouched.

However, the fact that this is **so** simple reduces its value, as anyone could easily write that `replace(...).into_iter()` already. A `drain(range)` or `drain(start, end)` would be more compelling, and not so trivial either.

Reply:
`drain(start, end)` seems an odd alternative to me - even Vec and VecDeque have a `RangeBounds` argument, while they don't have a `range` members.

I do have my own odd alternative though: as a member of `RangeMut`, but that simply cannot adjust length. So rather, a member of `IterMut`, but that can't be initialized with bounds. Oh boy, this is hard...